distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Text_Sub_Range.enso

```diff
@@ -0,0 +1,123 @@
+           Arguments:
+           - text: Text to search
+           - predicate: Function to test each character, receives:
+             - index: current index
+               start: index the char array to start of grapheme cluster
+               end: index the char array to start of next grapheme cluster
+             return True to exit loop
+           Returns: either a Pair of char indices for current grapheme cluster or
+             Pair -1 (char array length) if not found.
+        find_sub_range_end = text->predicate->
+            iterator = BreakIterator.getCharacterInstance
+            iterator.setText text
+
+            loop index start end =
+                if end == -1 then (Pair -1 start) else
+                    if predicate index start end then (Pair start end) else
+                        @Tail_Call loop (index + 1) end iterator.next
+
+            loop 0 0 iterator.next
+
+        case this of
+            First count ->
+                if count <= 0 then (Range 0 0) else
+                    indices = find_sub_range_end text (index->_->_-> index+1 == count)
+                    Range 0 indices.second
+            Last count ->
+                if count <= 0 then (Range 0 0) else
+                    first_count = text.length - count
+                    indices = find_sub_range_end text (index->_->_-> index+1 == first_count)
+                    if indices.first == -1 then (Range 0 indices.second) else
+                        (Range indices.second (Text_Utils.char_length text))
+            Before delimiter ->
+                if delimiter.is_empty then (Range 0 0) else
+                    index = Text_Utils.index_of text delimiter
+                    if index == -1 then (Range 0 (Text_Utils.char_length text)) else
+                        (Range 0 index)
+            Before_Last delimiter ->
+                if delimiter.is_empty then (Range 0 0) else
+                    index = Text_Utils.last_index_of text delimiter
+                    if index == -1 then (Range 0 (Text_Utils.char_length text)) else
+                        (Range 0 index)
+            After delimiter ->
+                if delimiter.is_empty then (Range 0 0) else
+                    index = Text_Utils.index_of text delimiter
+                    if index == -1 then (Range 0 0) else
+                        (Range (index + Text_Utils.char_length delimiter) (Text_Utils.char_length text))
+            After_Last delimiter ->
+                if delimiter.is_empty then (Range 0 0) else
+                    index = Text_Utils.last_index_of text delimiter
+                    if index == -1 then (Range 0 0) else
+                        (Range (index + Text_Utils.char_length delimiter) (Text_Utils.char_length text))
+            While predicate ->
+                wrapped start end = predicate (Text_Utils.substring text start end) . not
+                indices = find_sub_range_end text (_->wrapped)
```

Comment:
That's good but wouldn't that be clearer?
```suggestion
                indices = find_sub_range_end text _-> start-> end->
                    predicate (Text_Utils.substring text start end) . not
```

Also seems like less indirection.